doc/users-guide/base/constParam.rst

```diff
@@ -0,0 +1,193 @@
+:title: Users Guide: Const/Param Declarations
+
+.. _ug-constParam:
+
+Constant Declarations
+=====================
+
+In addition to variables, Chapel supports the declaration of
+*constants* which can be thought of as variables whose values cannot
+be changed once they've been initialized.  Chapel supports two
+flavors of constants, each of which is introduced by a distinct
+keyword:
+
+* ``const`` introduces an execution-time constant whose value need not
+  be known to the compiler
+
+* ``param`` introduces a compile-time constant whose value must be
+  known to the compiler
+
+Declaring immutable variables as ``const`` or ``param`` provides
+helpful semantic information, both to human readers of Chapel
+programs, and to the compiler itself.  In practice, ``param`` values
+are most often used in cases where compile-time operations or
+specialization will reduce execution-time computation, resulting in
+better performance.  In other cases, certain language constructs are
+required to be ``param`` values for the sake of static typing or
+performance.  Examples include the index values of a heterogeneous
+tuple or the number of dimensions in a rectangular array.
+
+
+Declaring *consts*
+------------------
+
+Constant declarations are a lot like :ref:`variable declarations
+<ug-vars>`, except for the change in keyword.  As a simple example,
+the following statement declares an integer constant named *two* whose
+value is 2 (as you'd hope):
+
+.. literalinclude:: examples/users-guide/base/constParam.chpl
+  :caption:
+  :language: chapel
+  :lines: 3
+
+Unlike variables, attempting to assign to a constant after it has been
+initialized will result in a compile-time error.  For example, if we
+later attempted to perform this assignment:
+
+.. literalinclude:: examples/users-guide/base/constParam.chpl
+  :language: chapel
+  :lines: 6
+  :dedent: 2
```

Comment:
The way the test is currently written, it is needed/useful in order to get rid of the leading whitespace on the line in question (since we don't want the conditional showing up in the documentation -- it's just there to permit error cases to be tested separately from "correct" code).

The other way to handle this would be to not indent the line in the test code in the normal way, but I've generally been formatting the test code the natural way and using dedents to get rid of whitespace in the generated documentation files.